### ISSUE TITLE
Unexpected NullReferenceException when running application on Linux

### ISSUE THREAD

**Issue by Skyppid** [dotnet/runtime#94847]
### Description

We just upgraded our ASP app to .NET 8 to benefit from the performance improvements and the new C#12 features. Everything seemed to work fine until someone noticed that on Staging we had an issue loading our projects.

The issue occurs when trying to map the projects into DTOs. A *NullReferenceException* gets thrown.

> System.NullReferenceException: Object reference not set to an instance of an object.
>    at ProjectOverviewDto (ArrayClosure, ProjectModel)
>    at TDestination Mapster.TypeAdapterBuilder<TSource>.Map<TDestination>()
>    at TDestination Mapster.TypeAdapterBuilder<TSource>.AdaptToType<TDestination>()
>    at async Task<TDestination> Mapster.TypeAdapterExtensions.AdaptToTypeAsync<TDestination>(IAdapterBuilder builder)
>    at async Task<PaginatedList<TProject>> Common.Utilities.PaginatedListExtensions.Project<TSource, TProject>(PaginatedList<TSource> source, IMapper mapper)

Usually I would search for issues in our codebase or try with Mapster - but in this peculiar case it's related to .NET 8 vs. .NET 7. The main difference is that our staging cluster runs Linux while we develop locally on Windows. When I redeploy the old .NET 7 version (with absolutely no changes to the code besides upgrading .NET) it starts to work again.

I tried to reproduce the issue by cloning the staging environment locally (so same codebase and same database) and it just works with .NET 8. But not on Staging.

So my assumption is: .NET 8 has some weird runtime issue on Linux.

The question is now: How can I provide you any help in finding it because I can't really tell how I could possibly help you find the issue.

### Reproduction Steps

Uncertain how you could reproduce this. I can't just give you the whole source code to test with. Maybe I can provide you with the IL generated, at least the compiled image or so. I don't know.

### Expected behavior

The state of the application was absolutely fine. No bugs whatsoever. So the upgrade should have not introduced weird Heisenbugs.

### Actual behavior

Working code started to fail with the exception above.

### Regression?

The last .NET 7 version works perfectly fine. Once we reverted the upgrade it started working again.

### Configuration

Working:
* .NET 7.0.310
* Windows (local) / Linux Debian [20.04.6 LTS (Focal Fossa)] (Cloud) - using the official image (`mcr.microsoft.com/dotnet/aspnet:7.0`)
* The following settings are applied in the Dockerfile:
```
ENV DOTNET_TieredPGO 1 
ENV DOTNET_TC_QuickJitForLoops 1 
ENV DOTNET_ReadyToRun 0
```
* Architecture x64 (local and Cloud)

Not working:
* .NET 8.0.100
* Same as above

**Comment by jakobbotsch** [comment]
Does setting `DOTNET_JitEnablePhysicalPromotion=0` help? What about `DOTNET_JITMinOpts=1`?

Is it possible for you to share a crash dump from when the exception happens?

**Comment by jakobbotsch** [comment]
If the exception is unhandled (i.e. it results in a crash), then you can set environment variables to automatically create a core dump when it happens: https://learn.microsoft.com/en-us/dotnet/core/diagnostics/collect-dumps-crash

**Comment by Skyppid** [comment]
It will be handled by the ASP middleware at latest I guess. Or does this apply to first-level exceptions? It's not inside a try-catch but somewhere along the stack it'll be handled, so the app won't crash completely.

**Comment by Skyppid** [comment]
> Does setting `DOTNET_JitEnablePhysicalPromotion=0` help? What about `DOTNET_JITMinOpts=1`?
> 
> Is it possible for you to share a crash dump from when the exception happens?

I can verify that setting `DOTNET_JitEnablePhysicalPromotion=0` indeed solves the issue! I reverted the change and it showed the same error again.

**Comment by jakobbotsch** [comment]
> It will be handled by the ASP middleware at latest I guess. Or does this apply to first-level exceptions? It's not inside a try-catch but somewhere along the stack it'll be handled, so the app won't crash completely.
> 
> Is it possible to create a mini dump by code? I could trigger it when it occurs and might be able to get the dump from the pod then. As of now I can't reproduce it locally.

I think you can call `Environment.FailFast` when you observe the situation happening to force a crash that will produce a crash dump. I also think that ASP.NET might know about the environment variables I linked, so you may just see if a crash dump is produced without doing anything when you set those environment variables (I found some information here: https://learn.microsoft.com/en-us/troubleshoot/developer/webapps/aspnetcore/practice-troubleshoot-linux/lab-1-3-capture-core-crash-dumps)

> I can verify that setting DOTNET_JitEnablePhysicalPromotion=0 indeed solves the issue! I reverted the change and it showed the same error again.
> 
> EDIT: Now it stopped working again with it being enabled... Guess it's back to .NET 7 then...

Hmm, so to clarify, you were still able to reproduce the issue with `DOTNET_JitEnablePhysicalPromotion=0`? What about `DOTNET_JITMinOpts=1`?

**Comment by Skyppid** [comment]
I'll try to create a dump with failfast.
I'm just not sure how I can get the dump since the container will be immediately restarted when it crashes.

Should I use both flags at the same time or only one of both? Yeah I think I might have been to fast and the container was not yet running fully. So likely it was still the .NET 7 backend that responded when it worked.

**Comment by jakobbotsch** [comment]
> I'll try to create a dump with failfast. I'm just not sure how I can get the dump since the container will be immediately restarted when it crashes.
> 
> Should I use both flags at the same time or only one of both? Yeah I think I might have been to fast and the container was not yet running fully. So likely it was still the .NET 7 backend that responded when it worked.
> 
> **EDIT**: JITMinOpts did not do anything, except for seemingly reduce performance? At least it took significantly longer the first time until the error appeared. Yet same problem.

Yeah, `DOTNET_JITMinOpts=1` will cause the JIT not to optimize any functions, so reduced performance is expected. If it still reproduces when running in that mode it usually means that this is not a JIT bug.

I'm afraid I'm not knowledgeable enough on the ASP.NET/container side to be able to help you figure out how to save the dump... perhaps there is some possibility of mounting a volume where the dump can be persisted?

**Comment by Skyppid** [comment]
Well me and my colleague were able to retrieve a dump. Is there some service you use where I can provide you with the dump?

**Comment by jakobbotsch** [comment]
> Well me and my colleague were able to retrieve a dump. Is there some service you use where I can provide you with the dump?

Do you have some file sharing service like OneDrive available where you can put it? Can you reach out to me over email? My email is janielsen at the company I work for.

**Comment by Skyppid** [comment]
I sent you an email with the OneDrive link.

**Comment by JulieLeeMSFT** [comment]
Ping @jakobbotsch, is there any progress on this investigation?

**Comment by jakobbotsch** [comment]
Sorry, this fell off my radar.
I've tried to look at the crashdump that was shared with me, but it appears to be incomplete. `dotnet-symbol --verifycore` complains with
```
ERROR: Core file not complete: file size 0x0051B000 segments total 0x033D9000
```
so it seems like the crash dump would be expected to be a lot larger than 5 MB.

@Skyppid If possible, can you please collect a new crash dump? Also, I'd suggest trying to track down where the `null` is coming from and why it is unexpected to be null -- so far, I don't see signs of a JIT bug here.

**Comment by Skyppid** [comment]
I just came back from my vacation and have to catch up on things. I'll check on this the coming days. But how comes the dump is incomplete? We followed the guidelines on Learn. It's a bit difficult to trigger in the right moment and also from remote in a AKS Pod. Not sure why it would be incomplete...

**Comment by jakobbotsch** [comment]
Thanks @Skyppid! I'm not sure why the dump would be incomplete. I wonder if perhaps the container restarted while the coredump was being captured, or if perhaps there was a problem storing the full coredump. Sorry to be unhelpful -- this is not my area of expertise.

**Comment by jakobbotsch** [comment]
@Skyppid https://github.com/dotnet/runtime/issues/91571#issuecomment-1707088139 has some information that might be helpful.

**Comment by Skyppid** [comment]
So we ran some tests on an extra environment and we found out something strange:

Currently our Dockerfile runs the app with this:
```
RUN dotnet publish "Web/Web.csproj" -c Release -o /app/publish

FROM base AS final
COPY --from=build /app/publish .
ENTRYPOINT ["dotnet", "App.Web.dll"]
```

If we modify it to `ENTRYPOINT["dotnet","run","--project","Web"]` it works. So maybe this is not an issue with your area, but something else. Seems like deploy actually breaks something.

So maybe you can get someone involved who can help here? I guess it might not be JIT after all.

**Comment by jkotas** [comment]
`dotnet run` uses Debug configuration by default. 

Can you try `dotnet publish "Web/Web.csproj" -c Debug -o /app/publish`? It will publish the app with the same optimization settings as `dotnet run` .

**Comment by Skyppid** [comment]
@jkotas We tried it and it did not work. Seems there is something off. 

We tried generating a dump when the exception gets caught, but it fails saying that the runtime does not match for the process. I assume `Process.GetCurrentProcess().Id` should deliver the correct process Id though. My colleague will check the versions tomorrow but I'm quite sure we used the latest 8.0 SDK and dotnet-dump tool.
```
            var dumpPath = OperatingSystem.IsLinux() ? "/tmp/dump.dmp" : Path.Combine(Environment.CurrentDirectory, "dump.dmp");

            _logger.LogInformation("Generating dump ...");
            ProcessStartInfo info =
                new("dotnet-dump", $"collect -p {Process.GetCurrentProcess().Id} --type Full -o {dumpPath}");
            var process = Process.Start(info);
            process!.ErrorDataReceived += (sender, args) => _logger.LogError(args.Data);
            process.OutputDataReceived += (sender, args) => _logger.LogInformation(args.Data);

            process!.WaitForExit(20000);
            _logger.LogInformation("Dump generated under {Path}.", dumpPath);
```

**Comment by Skyppid** [comment]
After hours of searching, trial & error and eliminating causes we ended up destroying even our .NET 7 app. Turns out that the update for [FastExpressionCompiler](https://github.com/dadhi/FastExpressionCompiler) seems to have a severe issue on it's end. Completely forgot we were using it but made sense that there might be the issue after downgrading to .NET 7 still didn't fix it.

So I'll close this as it's not a .NET issue (would have been surprised if it was). Thanks for your help though!
